### DISCUSSION TITLE
.mpy dynamic module for armv7-a

### DISCUSSION THREAD

**Discussion by ozgurgulderen** [micropython/micropython#16258] [Unix & Windows]
Hi,

I am trying to create a .mpy dynamic module for use with my AM335x based device running linux and need support for armv7a. I tried to extend the dynruntime.mk and mpy_ld.py to extend the supported architectures but have no luck to create a .mpy. What can be done to create dynamic modules for ARM Cortex-A8 ? 

Is there a way to do this?

I also tried to create the same dynamic module for armv6 but no luck. Linking fails always for libgcc relevanr symbols .e.g. __aeabi_uidiv

**Comment by jonnor** [comment]
In current git master, dynamic modules are not linked against libc. Which is why you get that symbol error. This PR fixes that, https://github.com/micropython/micropython/pull/15838

**Reply by ozgurgulderen** [reply]
Thanks a lot, this PR solves the issues for libgcc and libm. However, my dynamic module make process reports:


```
LinkError: /usr/arm-linux-gnueabihf/lib/libm.a:math_errf.o: undefined symbol: errno
/usr/lib/gcc-cross/arm-linux-gnueabihf/11/libgcc.a:_dvmd_lnx.o: undefined symbol: raise
make: *** [../micropython/py/dynruntime.mk:166: build/mymodule.native.mpy] Error 1

```

In dynruntime.mk, I have the following extension for armv7-a

```
else ifeq ($(ARCH),armv7a)
CROSS = arm-linux-gnueabihf-
MICROPY_ARCH_CFLAGS += -march=armv7-a -mfpu=neon -mfloat-abi=hard -fno-stack-protector -fno-builtin
MICROPY_FLOAT_IMPL ?= double
```


I am suspecting that as I am using arm-linux-gnueabihf cross compiler for Linux target, this requires some Linux OS support as the undefined reference of the raise Linux system call indicates.

I am not sure how to do native glue. 

PS: I crudely tried following extention to PR which is WRONG but gave it a shot

```
ifeq ($(MICROPY_LINK_RUNTIME),1)
  LIBGCC_PATH := $(realpath $(shell $(CROSS)gcc $(CFLAGS) --print-libgcc-file-name))
  LIBC_PATH := $(realpath $(shell $(CROSS)gcc $(CFLAGS) --print-file-name=libc.a))
  LIBM_PATH := $(realpath $(shell $(CROSS)gcc $(CFLAGS) --print-file-name=libm.a))
  MPY_LD_FLAGS += $(addprefix -l, $(LIBGCC_PATH) $(LIBM_PATH) $(LIBC_PATH))
endif
```
The output was:

```
LinkError: /usr/arm-linux-gnueabihf/lib/libc.a:dl-support.o: .data non-empty
make: *** [../micropython/py/dynruntime.mk:164: build/desigoxlib.native.mpy] Error 1
```

The error seems to be caused by attempting to link against a static libc in a way that is not compatible with the MicroPython dynamic runtime module build system. The specific error "dl-support.o: .data non-empty" suggests that the static libc object file contains initialized data sections, which are not allowed in dynamic runtime modules.

**Reply by jonnor** [reply]
Hmm. What is actually the code in your dynamic module? Have you tried a module that is trivial, like just a function returning a value, or one of the examples? Dynamic native modules do work on x64 on Linux when using the Unix port. Intuitively, I would thiiink that the syscall thing would be similar on x64 Linux to ARM Linux?
